File: credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    //////// get cardnumber from prompt ////////
    long cardnumber;
    do
    {
        cardnumber = get_long("Number: ");
    }
    while (cardnumber < 0);
    
    /////// calculate checksum ///////
    long checksum = 0;
    int length = 0;
    int first2digit = 0;
    long cnd = 0;  /// card number second-to-last digit 
    long cndf = 0; /// card number first-to-last digit 
    
    for (long d = 100; d < cardnumber*10; d *= 100) //iterate start from the second-to-last digit
    {
        cnd = ((cardnumber%d - cardnumber%(d/10)) / (d/10))*2;
        length += 1;
        /// printf("Digit: %li\n", cnd);
        if (cnd >= 10)
        {
            cnd = (cnd%100 - cnd%10) / 10 + (cnd%10);
            checksum += cnd;
        }
        else if (cnd < 10 )
        {
            checksum += cnd;
        }
        
    }
    /// printf("%li\n\n", checksum);
    
    for (long df = 10; df < cardnumber*10; df *= 100) //start fro the first-to-last digit
    {
        length += 1;
        cndf = ((cardnumber%df - cardnumber%(df/10)) / (df/10));
        /// printf("cndf= %li\n", cndf);
        checksum += cndf;
        /// printf("%li\n", checksum);
    }
    ///printf("Length: %i\n", length); ////// get lentgh plus one after evert loop
 
    ////// Check for card lentgh to catagorize wich card belong and validate accordingly ///////
    if (checksum % 10 == 0) /// valid number
    {
        if (length % 2 == 0)
        {
            first2digit += cnd/2 * 10 + cndf;
        }
        else if (length % 2 == 1)
        {
            first2digit += cndf * 10 + cnd;
        }
        
        ///printf ("%i\n", first2digit);
        
        if (length == 15 && (first2digit == 34 || first2digit == 37))
        {
            printf("AMEX\n");
        }
        else if (length == 16 && 51 <= first2digit && first2digit <= 55)
        {
            printf("MASTERCARD\n");
        }
        else if ((length == 13 || length == 16) && (first2digit - (first2digit % 10))/10  == 4)
        {
            printf("VISA\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}


void checksum(void)
{
    
}

File: credit.py
import cs50


def main():
    n = number()
    sum = calsd(n) + calfd(n)
    checksum(n, sum, len(str(n)))
    
# get cardnumber from prompt
def number():
    while True:
        n = cs50.get_int("Number: ")
        if n > 0:
            break
    return n
    
# even digits sum (count from second-to-last digit)
def calsd(n):
    sdsum = 0
    for i in range(1, int(len(str(n))/2) + 1):
        d = 100 ** i
        sd = (n % d - n % (d/10)) / (d/10)
        sd *= 2
        if sd >= 10:
            sd = (sd % 100 - sd % 10) / 10 + (sd % 10)
        sdsum += sd
    return sdsum
        
    
# odd digits sum (count from first-to-last digit)
def calfd(n):
    fdsum = 0
    for i in range(int(len(str((n)))/2 + 1)):
        d = 10 * (100 ** i)
        fd = (n % d - n % (d/10)) / (d/10)
        fdsum += fd
    return fdsum

# checksum
def checksum(n, sum, len):
    f2d = (n - n % (10 ** (len-2))) / (10 ** (len-2))  # first 2 digit of cardnumber
    
    if int(sum) % 10 == 0:
        if (f2d == 34 or f2d == 37 ) and len == 15:
            print("AMEX")
        elif 51 <= f2d <= 55 and len == 16:
            print("MASTERCARD")
        elif (f2d - f2d % 10) / 10 == 4 and (len == 13 or len == 16):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")
    
    
main()